File: python/src/wh40kdc/export/atc_2026.py
"""ATC 2026 roster text exporters (``atc-2026-compact`` / ``atc-2026-full``).

These reuse the WTC compact/full *bodies* verbatim (see
:func:`~wh40kdc.export.newrecruit_wtc.wtc_compact_body_lines` /
:func:`~wh40kdc.export.newrecruit_wtc.wtc_full_body_lines`) but replace the
summary header with the block the American Team Championship 2026 list-
submission format asks for: player/team identification, the picked Force
Disposition, every enhancement-bearing model, and the leader/support
attachments spelled out.

Provisional and **export-only** — there is no ATC import adapter, so the format
is additive. The existing WTC formats (and real-world WTC import) are untouched.

Python mirror of ``tools/src/export/atc-2026.ts``; byte-identical output.
"""

from __future__ import annotations

from wh40kdc.export.helpers import (
    Roster,
    RosterUnit,
    char_slot_assignment,
    title_case_id,
    total_army_points,
)
from wh40kdc.export.newrecruit_wtc import (
    _FENCE,
    wtc_compact_body_lines,
    wtc_full_body_lines,
)

_DASH = "—"


def _header(roster: Roster, units: list[RosterUnit], char_slots: list[int | None]) -> str:
    faction = title_case_id(roster.get("faction_id")) or "Unknown"
    disposition = title_case_id(roster.get("force_disposition")) or _DASH
    detachments = roster["detachments"]
    detachment = (
        ", ".join(title_case_id(d["ref"]["id"]) or d["ref"]["raw_name"] for d in detachments)
        if detachments
        else _DASH
    )
    total = roster["points"].get("total_reported")
    if total is None:
        total = total_army_points(roster)

    warlord = _DASH
    for i, u in enumerate(units):
        if u.get("is_warlord"):
            warlord = f"Char{char_slots[i]}: {u['ref']['raw_name']}"
            break

    enh_parts = [
        f"{u['enhancement']['raw_name']} (on Char{char_slots[i]}: {u['ref']['raw_name']})"
        for i, u in enumerate(units)
        if u.get("enhancement") is not None
    ]
    enhancement = "; ".join(enh_parts) if enh_parts else _DASH

    attach_parts = [
        f"{u['ref']['raw_name']} attached to {u['leader_attachment']['bodyguard_ref']['raw_name']}"
        for u in units
        if u.get("leader_attachment") is not None
    ]
    leader_support = "; ".join(attach_parts) if attach_parts else _DASH

    lines = [
        _FENCE,
        f"+ PLAYER NAME: {_DASH}",
        f"+ TEAM NAME: {_DASH}",
        f"+ FACTIONS USED: {faction}",
        f"+ DISPOSITION: {disposition}",
        f"+ DETACHMENT: {detachment}",
        f"+ ARMY POINTS: {total}pts",
        "+",
        f"+ WARLORD: {warlord}",
        f"+ ENHANCEMENT: {enhancement}",
        f"+ LEADER/SUPPORT: {leader_support}",
        f"+ NUMBER OF UNITS: {len(units)}",
        _FENCE,
    ]
    return "\n".join(lines)


def serialize_atc_2026_compact(roster: Roster) -> str:
    units = roster["units"]
    slots = char_slot_assignment(units)
    return "\n".join([_header(roster, units, slots), *wtc_compact_body_lines(units, slots)]) + "\n"


def serialize_atc_2026_full(roster: Roster) -> str:
    units = roster["units"]
    slots = char_slot_assignment(units)
    return "\n".join([_header(roster, units, slots), *wtc_full_body_lines(units, slots)])

File: crates/wh40kdc/src/export/atc_2026.rs
//! ATC 2026 roster text exporters (`atc-2026-compact` / `atc-2026-full`).
//!
//! These reuse the WTC compact/full *bodies* verbatim (see
//! [`wtc_compact_body_lines`] / [`wtc_full_body_lines`]) but replace the
//! summary header with the block the American Team Championship 2026 list-
//! submission format asks for: player/team identification, the picked Force
//! Disposition, every enhancement-bearing model, and the leader/support
//! attachments spelled out.
//!
//! Provisional and **export-only** — there is no ATC import adapter, so the
//! format is additive. The existing WTC formats (and real-world WTC import)
//! are untouched.
//!
//! Rust mirror of `tools/src/export/atc-2026.ts`.

use crate::import::{Roster, RosterUnit};

use super::helpers::{char_slot_assignment, title_case_id, total_army_points};
use super::newrecruit_wtc::{wtc_compact_body_lines, wtc_full_body_lines, FENCE};
use super::{ExportFormat, RosterSerializer};

const DASH: &str = "—";

fn atc_header(roster: &Roster, units: &[RosterUnit], char_slots: &[Option<u32>]) -> String {
    let faction = title_case_id(roster.faction_id.as_deref()).unwrap_or_else(|| "Unknown".into());
    let disposition =
        title_case_id(roster.force_disposition.as_deref()).unwrap_or_else(|| DASH.into());
    let detachment = if roster.detachments.is_empty() {
        DASH.to_string()
    } else {
        roster
            .detachments
            .iter()
            .map(|d| title_case_id(d.ref_.id.as_deref()).unwrap_or_else(|| d.ref_.raw_name.clone()))
            .collect::<Vec<_>>()
            .join(", ")
    };
    let total = roster
        .points
        .total_reported
        .unwrap_or_else(|| total_army_points(roster));

    let warlord = match units.iter().position(|u| u.is_warlord) {
        Some(i) => format!(
            "Char{}: {}",
            char_slots[i].map(|n| n.to_string()).unwrap_or_default(),
            units[i].ref_.raw_name
        ),
        None => DASH.to_string(),
    };

    let enh_parts: Vec<String> = units
        .iter()
        .enumerate()
        .filter_map(|(i, u)| {
            u.enhancement.as_ref().map(|enh| {
                format!(
                    "{} (on Char{}: {})",
                    enh.raw_name,
                    char_slots[i].map(|n| n.to_string()).unwrap_or_default(),
                    u.ref_.raw_name
                )
            })
        })
        .collect();
    let enhancement = if enh_parts.is_empty() {
        DASH.to_string()
    } else {
        enh_parts.join("; ")
    };

    let attach_parts: Vec<String> = units
        .iter()
        .filter_map(|u| {
            u.leader_attachment.as_ref().map(|la| {
                format!(
                    "{} attached to {}",
                    u.ref_.raw_name, la.bodyguard_ref.raw_name
                )
            })
        })
        .collect();
    let leader_support = if attach_parts.is_empty() {
        DASH.to_string()
    } else {
        attach_parts.join("; ")
    };

    let lines = vec![
        FENCE.to_string(),
        format!("+ PLAYER NAME: {DASH}"),
        format!("+ TEAM NAME: {DASH}"),
        format!("+ FACTIONS USED: {faction}"),
        format!("+ DISPOSITION: {disposition}"),
        format!("+ DETACHMENT: {detachment}"),
        format!("+ ARMY POINTS: {total}pts"),
        "+".to_string(),
        format!("+ WARLORD: {warlord}"),
        format!("+ ENHANCEMENT: {enhancement}"),
        format!("+ LEADER/SUPPORT: {leader_support}"),
        format!("+ NUMBER OF UNITS: {}", units.len()),
        FENCE.to_string(),
    ];
    lines.join("\n")
}

pub struct Atc2026CompactSerializer;

impl RosterSerializer for Atc2026CompactSerializer {
    fn id(&self) -> ExportFormat {
        ExportFormat::Atc2026Compact
    }

    fn serialize(&self, roster: &Roster) -> String {
        let units = &roster.units;
        let slots = char_slot_assignment(units);
        let mut lines: Vec<String> = vec![atc_header(roster, units, &slots)];
        lines.extend(wtc_compact_body_lines(units, &slots));
        let mut out = lines.join("\n");
        out.push('\n');
        out
    }
}

pub struct Atc2026FullSerializer;

impl RosterSerializer for Atc2026FullSerializer {
    fn id(&self) -> ExportFormat {
        ExportFormat::Atc2026Full
    }

    fn serialize(&self, roster: &Roster) -> String {
        let units = &roster.units;
        let slots = char_slot_assignment(units);
        let mut lines: Vec<String> = vec![atc_header(roster, units, &slots)];
        lines.extend(wtc_full_body_lines(
            units,
            &slots,
            roster.faction_id.as_deref(),
        ));
        lines.join("\n")
    }
}
